src/uct/cuda/cuda_ipc/cuda_ipc_ep.c

```diff
@@ -22,19 +22,47 @@
 #define UCT_CUDA_IPC_PUT 0
 #define UCT_CUDA_IPC_GET 1
 
+
+static UCS_F_ALWAYS_INLINE void
+uct_cuda_ipc_ep_set_keepalive_device(uct_cuda_ipc_ep_t *ep)
+{
+    if ((ep->keepalive.device == CU_DEVICE_INVALID) &&
+        (UCT_CUDADRV_FUNC(cuCtxGetDevice(&ep->keepalive.device),
+                          UCS_LOG_LEVEL_TRACE) != UCS_OK)) {
+        /* assume CUDA-IPC transport won't be used, if it fails to get device
+         * context. If this EP will be used then, cudaSetDevice() will be
+         * invoked, and then keepalive.device will be set */
+        ep->keepalive.device = CU_DEVICE_INVALID;
+    }
+}
+
+
 static UCS_CLASS_INIT_FUNC(uct_cuda_ipc_ep_t, const uct_ep_params_t *params)
 {
     uct_cuda_ipc_iface_t *iface = ucs_derived_of(params->iface,
                                                  uct_cuda_ipc_iface_t);
+    const uct_cuda_ipc_iface_addr_t *iface_addr;
 
     UCT_EP_PARAMS_CHECK_DEV_IFACE_ADDRS(params);
     UCS_CLASS_CALL_SUPER_INIT(uct_base_ep_t, &iface->super);
 
+    iface_addr = (const uct_cuda_ipc_iface_addr_t*)params->iface_addr;
+
+    self->keepalive.device = CU_DEVICE_INVALID;
+    uct_cuda_ipc_ep_set_keepalive_device(self);
+
```

Comment:
Should we throw an error here if no device is found at the time of creating ep?

Reply:
but then many applications will fail, e.g. if an application wants to use a host memory buffer, but we create CUDA-IPC EP anyway.

Reply:
I assumed we create endpoints lazily and defer cuda_ipc ep creation only until a large message device to device transfer is needed.